### ISSUE TITLE
[Windows] fluent-bit agent exit 1 after out_stackdriver cannot connect to metadata server 

### ISSUE THREAD

**Issue by sophieyfang** [fluent/fluent-bit#2880]
## Bug Report

**Describe the bug**
This a bug is introduced between 1.6.7 and 1.6.8. v1.6.7 works just as expected and with no trouble, while v1.6.8 has trouble connecting to the metadata server and the agent exit 1 right afterwards.

Under the condition they use the same configuration file:
"""
[Service]
    Log_Level    debug
[INPUT]
    Name         winlog
    Channels     Setup,Windows PowerShell
    Interval_Sec 1
    DB           winlog.sqlite
[OUTPUT]
    Name         stackdriver
    resource     gce_instance
    Match        *
    Retry_Limit  3
"""
By running the v1.6.3 in the console, the console logs show it can connect to the metadata server just fine: 
![image](https://user-images.githubusercontent.com/8354694/102718868-9c61c180-42b8-11eb-82ef-66776d480af6.png)

Versus By running the v1.6.8 in the console, the console logs show it cannot connect to the metadata server and the agent exits right afterwards:
![image](https://user-images.githubusercontent.com/8354694/102718879-a5529300-42b8-11eb-81dc-aaa6e8f38939.png)

**To Reproduce**
- Download v1.6.7 and v1.6.8

   - Inside the Windows Powershell:

   ```
    $url="https://fluentbit.io/releases/1.6/td-agent-bit-1.6.7-win64.exe"
    Invoke-WebRequest -Uri $url -OutFile .\td-agent-bit-1.6.7-win64.exe
    $url="https://fluentbit.io/releases/1.6/td-agent-bit-1.6.8-win64.exe"
    Invoke-WebRequest -Uri $url -OutFile .\td-agent-bit-1.6.8-win64.exe
   ```
   - Run the agent in Windows Powershell Admin mode:
   
   ```
   & 'C:\Program Files\td-agent-bit-1.6.7\bin\fluent-bit.exe' -vvv --config='C:\tmp\fluent-bit.conf'
   & 'C:\Program Files\td-agent-bit-1.6.8\bin\fluent-bit.exe' -vvv --config='C:\tmp\fluent-bit.conf'
   ```
- Example error log message in v1.6.8:
```
[error] [io] connection #624 failed to: metadata.google.internal:80
[debug] [upstream] connection #624 failed to metadata.google.internal:80
[error] [output:stackdriver:stackdriver.0] failed to create metadata connection
[error] [output:stackdriver:stackdriver.0] can't fetch token from the metadata server
[error] [lib] backend failed
```
**Expected behavior**
v1.6.7 works just fine and has the expected behavior.
```
[2020/12/17 19:52:39] [debug] [TLS] cannot parse a certificate. skipping...
[2020/12/17 19:52:39] [debug] [http_client] not using http_proxy for header
[2020/12/17 19:52:39] [debug] [http_client] header=GET /computeMetadata/v1/instance/service-accounts/default/token HTTP/1.1
Host: metadata.google.internal:80
Content-Length: 0
User-Agent: Fluent-Bit
Content-Type: application/text
Metadata-Flavor: Google

[2020/12/17 19:52:39] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/12/17 19:52:39] [debug] [upstream] KA connection #652 to metadata.google.internal:80 is now available
[2020/12/17 19:52:39] [debug] [upstream] KA connection #652 to metadata.google.internal:80 has been assigned (recycled)
[2020/12/17 19:52:39] [debug] [http_client] not using http_proxy for header
[2020/12/17 19:52:39] [debug] [http_client] header=GET /computeMetadata/v1/project/project-id HTTP/1.1
Host: metadata.google.internal:80
Content-Length: 0
User-Agent: Fluent-Bit
Content-Type: application/text
Metadata-Flavor: Google

[2020/12/17 19:52:39] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/12/17 19:52:39] [debug] [upstream] KA connection #652 to metadata.google.internal:80 is now available
[2020/12/17 19:52:39] [debug] [upstream] KA connection #652 to metadata.google.internal:80 has been assigned (recycled)
[2020/12/17 19:52:39] [debug] [http_client] not using http_proxy for header
[2020/12/17 19:52:39] [debug] [http_client] header=GET /computeMetadata/v1/instance/zone HTTP/1.1
Host: metadata.google.internal:80
Content-Length: 0
User-Agent: Fluent-Bit
Content-Type: application/text
Metadata-Flavor: Google

[2020/12/17 19:52:39] [debug] [output:stackdriver:stackdriver.0] HTTP Status=200
[2020/12/17 19:52:39] [debug] [upstream] KA connection #652 to metadata.google.internal:80 is now available
[2020/12/17 19:52:39] [debug] [upstream] KA connection #652 to metadata.google.internal:80 has been assigned (recycled)
[2020/12/17 19:52:39] [debug] [http_client] not using http_proxy for header
[2020/12/17 19:52:39] [debug] [http_client] header=GET /computeMetadata/v1/instance/id HTTP/1.1
Host: metadata.google.internal:80
Content-Length: 0
User-Agent: Fluent-Bit
Content-Type: application/text
Metadata-Flavor: Google
```

**Your Environment**
* Version used: v1.6.8
* Configuration:
```
[Service]
    Log_Level    debug
[INPUT]
    Name         winlog
    Channels     Setup,Windows PowerShell
    Interval_Sec 1
    DB           winlog.sqlite
[OUTPUT]
    Name         stackdriver
    resource     gce_instance
    Match        *
    Retry_Limit  3
```
* Environment name and version (e.g. Kubernetes? What version?): Windows Server 2016 Datacenter
* Server type and version: Windows Server 2016 Datacenter
* Operating System and version: Windows Server 2016 Datacenter
* Filters and plugins: See the configuration.

**Additional context**

out_stackdriver plugin is the key to send log to the Cloud Logging API. 

Something changed between 1.6.7 and 1.6.8 in the output plugin side or the library the output plugin uses for connection. [Here is the code comparison](https://github.com/fluent/fluent-bit/compare/v1.6.7...v1.6.8) between v1.6.8 and v1.6.7. After looking thru the code, `failed to create metadata connection` is one hint i saw in the error message, which leads to [this line](https://github.com/fluent/fluent-bit/blob/v1.6.8/plugins/out_stackdriver/gce_metadata.c#L64) and function flb_upstream_conn_get which later calls create_conn then emit error message `[upstream] connection #624 failed to metadata.google.internal:80` [in this line](https://github.com/fluent/fluent-bit/blob/v1.6.8/src/flb_upstream.c#L312). So the question led to the function [flb_io_net_connect](https://github.com/fluent/fluent-bit/commits/v1.6.8/src/flb_io.c).

**Comment by fujimotos** [comment]
I believe this is resolved by #2856.

**Comment by sophieyfang** [comment]
This is great!! It seems that the only problem left now is that I have
trouble compiling it from the source even following the instruction.
https://github.com/fluent/fluent-bit-docs/issues/427#issuecomment-748665161

**Comment by agup006** [comment]
@sophieyfang I believe that fix should be available in Fluent Bit 1.6.9
